pkg/builder/kaniko/kaniko.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	fmt.Printf("Cleaning up Job: %s\n", kJob.Name)
+
+	if err := k.cleanup(ctx, kJob); err != nil {
+		return "", fmt.Errorf("error cleaning up: %w", err)
+	}
+
+	fmt.Printf("Build completed successfully\n")
+
+	return logs, nil
+}
+
+func (k *Kaniko) waitForJobCompletion(ctx context.Context, job *batchv1.Job) (*batchv1.Job, error) {
+	watcher, err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", job.Name),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error watching Job: %w", err)
+	}
+	defer watcher.Stop()
+
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return nil, fmt.Errorf("watch channel closed unexpectedly")
+			}
+
+			j, ok := event.Object.(*batchv1.Job)
+			if !ok {
+				continue
+			}
+
+			if j.Status.Succeeded > 0 || j.Status.Failed > 0 {
+				// Job completed (successfully or failed)
+				return j, nil
+			}
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context cancelled")
+		}
+	}
+}
+
+func (k *Kaniko) getFirstPodFromJob(ctx context.Context, job *batchv1.Job) (*v1.Pod, error) {
+	// Assuming there's only one Pod template in the Job
+	podList, err := k.k8sClientset.CoreV1().Pods(k.k8sNamespace).List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("job-name=%s", job.Name),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing Pods: %w", err)
+	}
+
+	if len(podList.Items) == 0 {
+		return nil, fmt.Errorf("no Pods found for the Job: %s", job.Name)
+	}
+
+	return &podList.Items[0], nil
+}
+
+func (k *Kaniko) getContainerLogs(ctx context.Context, pod *v1.Pod) (string, error) {
+	if len(pod.Spec.Containers) == 0 {
+		return "", fmt.Errorf("no containers found in Pod: %s", pod.Name)
+	}
+
+	containerName := pod.Spec.Containers[0].Name
+
+	logOptions := v1.PodLogOptions{
+		Container: containerName,
+	}
+
+	req := k.k8sClientset.CoreV1().Pods(k.k8sNamespace).GetLogs(pod.Name, &logOptions)
+	logs, err := req.DoRaw(ctx)
+	if err != nil {
+		return "", fmt.Errorf("error retrieving container logs: %w", err)
+	}
+
+	return string(logs), nil
+}
+
+func (k *Kaniko) cleanup(ctx context.Context, job *batchv1.Job) error {
+	err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).
+		Delete(ctx, job.Name, metav1.DeleteOptions{
+			PropagationPolicy: &[]metav1.DeletionPropagation{metav1.DeletePropagationBackground}[0],
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Job: %w", err)
+	}
+
+	// Delete the associated Pods
+	err = k.k8sClientset.CoreV1().Pods(k.k8sNamespace).
+		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("job-name=%s", job.Name),
+		})
+	if err != nil {
+		return fmt.Errorf("error deleting Pods: %w", err)
+	}
+
+	return nil
+}
+
+func GetDefaultCacheOptions(buildContext string) (*builder.CacheOptions, error) {
+	if buildContext == "" {
+		return nil, fmt.Errorf("build context cannot be empty")
```

Comment:
I'm a fan of potentially defining an `errors.go` file and naming these kind of errors, so they can be explicitly referenced/inspected outside, especially if there are multiple implementations of the builder, ie:

builder/errors#ErrEmpyBuildContext

then each implementation and return similar errors